File: FindMinimumRotatedSortedArray.cs
//Algo: Binary Search
//TC:O(logN)
//SC:O(1) no extra space used

public class Solution {
    public int FindMin(int[] nums) {
        int low = 0;
        int high = nums.Length-1;
        
        if(nums[low]<=nums[high]){
                return nums[low];
            }
        while(low<=high){
            int mid = low + (high-low)/2;
             if(nums[low]<=nums[high]){
                return nums[low];
            }
            if((mid>0 && nums[mid]<nums[mid-1])){
                return nums[mid];
            }else if(nums[mid]>=nums[low]){
                low = mid+1;
            }else{
                high = mid-1;
            }
        }
        
        return 1;
    }
}

File: FindMinimumRotatedSortedArray.java
//TimeComplexity : O(log(n))
//SpaceComplexity : O(1)

class FindMinimumRotatedSortedArray {
    public int findMin(int[] nums) {
        if(nums.length ==1){
            return  nums[0];
        }
        if(nums.length  == 0){
            return -1;
        }

        int left = 0;
        int right = nums.length-1;
        //binary search
        while(left<=right){
            int midpoint = left + right-left/2;
            if(midpoint > 0  && nums[midpoint] < nums[midpoint-1]){
                return nums[midpoint];
            }else if(nums[left] <= nums[midpoint] &&  nums[right] < nums[midpoint]){
                left = midpoint+1;
            }
            else{
                right = midpoint -1;
            }

        }
        return nums[left];
    }
}